llms/mlx_lm/utils.py

```diff
@@ -91,10 +91,35 @@ def get_model_path(path_or_hf_repo: str) -> Path:
     return model_path
 
 
+def apply_repetition_penalty(logits: mx.array, generated_tokens: Any, penalty: float):
+    """
+    Apply repetition penalty to specific logits based on the given context.
+
+    Args:
+        logits (mx.array): The logits produced by the language model.
+        generated_tokens (any): A list of N previous tokens.
+        penalty (float): The repetition penalty factor to be applied.
+
+    Returns:
+        logits (mx.array): Logits with repetition penalty applied to generated tokens.
+    """
+    if len(generated_tokens) > 0:
+        indices = mx.array([token for token in generated_tokens])
+        selected_logits = logits[:, indices]
+        selected_logits = mx.where(
+            selected_logits < 0, selected_logits * penalty, selected_logits / penalty
+        )
+        logits[:, indices] = selected_logits
+    return logits
+
+
 def generate_step(
     prompt: mx.array,
     model: nn.Module,
-    temp: float = 0.0,
+    temp: 0.0,
+    repetition_penalty: Optional[float] = None,
+    repetition_context: Optional[List[int]] = None,
```

Comment:
(I thought I left this comment earlier, but it didn't save....)

Can repetition context simply be constructed inside the `generate_step` function rather passing it as a parameter? It looks like it is only used inside this function and you have all you need to construct it? Do we gain anything by passing it from outside the function?

Reply:
Maybe it was to apply the penalty for prompt, but the generate_step has the prompt tokens that can also be done inside the generate_step.

Reply:
Yes, I have removed `repetition_context` parameter,  and constructed it inside the `generate_step` function.

I have one suggestion, with the current approach we are applying the context size on the prompt as well. Maybe we apply repetition penalty for every single token in the prompt, and use context size only on the generated tokens?

Reply:
@vishal-14069 from what I can see in the transformers source code, it seems like the prompt included for it, but I cannot be 100% sure. Maybe you can double-check to see if it needs to be applied to the prompt.

https://github.com/huggingface/transformers/blob/main/src/transformers/generation/utils.py#L2645
https://github.com/huggingface/transformers/blob/main/src/transformers/generation/logits_process.py#L288

Reply:
Yes, the repetition penalty needs to be applied on the prompt as well. In my last comment what I was trying to say is that, in the latest approach we are using `repetition_context_size` on the prompt tokens.  

```python
repetition_context = []
     if repetition_penalty and repetition_penalty > 1.0:
         repetition_context = prompt_tokens.tolist()
         repetition_context = repetition_context[-repetition_context_size:] # this part 
```

I am suggesting that we use repetition penalty on the prompt tokens, but not the `repetition_context_size` on the prompt tokens, rather we apply repetition penalty on all the prompt tokens irrespective of the `repetition_context_size`  which will create a window on the prompt. 

Maybe we can use `repetition_context_size` only on the generated tokens. 

Suggested change:

```python
repetition_context = []
     if repetition_penalty and repetition_penalty > 1.0:
         repetition_context = prompt_tokens.tolist()
```

I can setup a PR, it will be easier for review.